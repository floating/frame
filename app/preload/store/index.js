/* globals fetch */

import { ipcRenderer } from 'electron'

import EventEmitter from 'events'
import Restore from 'react-restore'

import * as actions from './actions'
import state from './state'

import rpc from '../rpc'

import PersistStore from 'electron-store'

const persist = new PersistStore()

export const store = Restore.create(state(), actions)
store.events = new EventEmitter()

rpc('getSigners', (err, signers) => {
  if (err) return store.signersError(err)
  store.updateSigners(signers)
})

rpc('launchStatus', (err, status) => {
  if (err) return console.log(err) // launchStatusError
  store.setLaunch(status)
})

ipcRenderer.on('main:addSigner', (e, signer) => store.addSigner(signer))
ipcRenderer.on('main:removeSigner', (e, signer) => {
  if (store('signer.current') === signer.id) store.unsetSigner()
  store.removeSigner(signer)
})
ipcRenderer.on('main:updateSigner', (e, signer) => store.updateSigner(signer))
ipcRenderer.on('main:setSigner', (e, signer) => {
  if (signer.id) {
    store.setSigner(signer)
  } else {
    store.unsetSigner()
  }
})

const etherRates = () => {
  fetch('https://api.coinbase.com/v2/exchange-rates?currency=ETH').then(res => res.json()).then(res => {
    if (res && res.data && res.data.rates) store.updateExternalRates(res.data.rates)
  }).catch(e => console.log('Unable to fetch exchange rate', e))
}
etherRates()
setInterval(etherRates, 10000)

// Store Observers
let network = ''
store.observer(() => {
  if (network !== store('local.connection.network')) {
    network = store('local.connection.network')
    ipcRenderer.send('tray:setNetwork', network)
  }
})

store.observer(_ => persist.set('local', store('local')))

let launch = store('local.launch')
store.observer(() => {
  if (launch !== store('local.launch')) {
    launch = store('local.launch')
    if (launch) {
      rpc('launchEnable', err => console.log(err))
    } else {
      rpc('launchDisable', err => console.log(err))
    }
  }
})

export default store
