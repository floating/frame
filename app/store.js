/* globals */

import EventEmitter from 'events'
import Restore from 'react-restore'

import link from '../resources/link'

// const actions = {
//   initialSignerPos: (u, pos) => u('selected.position.initial', () => pos),
//   initialScrollPos: (u, pos) => u('selected.position.scrollTop', () => pos)
// }

import * as actions from '../resources/store/actions.panel'

export default (state, _cb) => {
  const store = Restore.create(state, actions)
  store.events = new EventEmitter()

  // Feed for relaying state updates
  // store.api.feed((state, actions, obscount) => {
  //   actions.forEach(action => {
  //     action.updates.forEach(update => {
  //       // console.log(update)
  //       // if (update.path.startsWith('main')) return
  //       // if (update.path.startsWith('panel')) return
  //       // link.send('tray:syncPath', update.path, update.value)
  //     })
  //   })
  // })

  link.on('action', (action, ...args) => {
    if (store[action]) store[action](...args)
  })

  return store
}
