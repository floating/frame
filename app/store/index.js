/* globals */

import EventEmitter from 'events'
import Restore from 'react-restore'

import link from '../../resources/link'
import * as actions from './actions'

export default (state, cb) => {
  const store = Restore.create(state, actions)
  store.events = new EventEmitter()

  // Feed for relaying state updates
  // store.api.feed((state, actions, obscount) => {
  //   actions.forEach(action => {
  //     action.updates.forEach(update => {
  //       // console.log(update)
  //       // if (update.path.startsWith('main')) return
  //       // if (update.path.startsWith('panel')) return
  //       // link.send('tray:syncPath', update.path, update.value)
  //     })
  //   })
  // })

  link.on('action', (action, ...args) => { if (store[action]) store[action](...args) })
  link.send('tray:ready') // turn on api

  link.send('tray:refreshMain')

  return store
}
