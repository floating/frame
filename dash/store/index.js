// import Restore from 'react-restore'
// import * as actions from './actions'
// export default (state) => Restore.create(state, actions)

/* globals fetch */

import EventEmitter from 'events'
import Restore from 'react-restore'

import link from '../../resources/link'
import * as actions from './actions'

export default (state, cb) => {
  // console.log('state...', state)
  const store = Restore.create(state, actions)
  store.events = new EventEmitter()
  link.on('action', (action, ...args) => {
    if (store[action]) store[action](...args)
  })

  return store
}
