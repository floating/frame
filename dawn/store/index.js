/* globals fetch */

import EventEmitter from 'events'
import Restore from 'react-restore'

import link from '../../resources/link'
import * as actions from './actions'

export default (state, cb) => {
  const store = Restore.create(state, actions)
  link.on('action', (action, ...args) => {
    if (store[action]) store[action](...args)
  })
  store.events = new EventEmitter()
  return store
}
