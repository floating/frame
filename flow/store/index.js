// import Restore from 'react-restore'
// import * as actions from './actions'
// export default (state) => Restore.create(state, actions)

/* globals fetch */

import EventEmitter from 'events'
import Restore from 'react-restore'

import link from '../../resources/link'
import * as actions from './actions'

export default (state, cb) => {
  const store = Restore.create(state, actions)
  store.events = new EventEmitter()

  // Feed for relaying state updates
  store.api.feed((state, actions, obscount) => {
    actions.forEach((action) => {
      action.updates.forEach((update) => {
        if (update.path.startsWith('main')) return
        link.send('tray:syncPath', update.path, update.value)
      })
    })
  })

  link.on('action', (action, ...args) => {
    if (store[action]) store[action](...args)
  })

  return store
}
