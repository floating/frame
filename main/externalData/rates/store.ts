import log from 'electron-log'
import { AssetType } from '@framelabs/pylon-client/dist/assetId'

import { storeApi } from '../storeApi'
import { toTokenId } from '../../../resources/domain/balance'

import type { AssetId } from '@framelabs/pylon-client/dist/assetId'
import type { Rate, WithTokenId } from '../../store/state/types'

type RateUpdate = {
  id: AssetId
  data: {
    usd: number
    usd_24h_change: number
  }
}

const RATES_EXPIRY = 1000 * 5 * 60 // rates are valid for 5 minutes before being considered stale
const timeouts: Record<string, NodeJS.Timer> = {}

const separateUpdates = (updates: RateUpdate[]) =>
  updates.reduce(
    (acc, update) => {
      return update.id.type === AssetType.NativeCurrency
        ? { ...acc, native: [...acc.native, update] }
        : { ...acc, token: [...acc.token, update] }
    },
    { native: [] as RateUpdate[], token: [] as RateUpdate[] }
  )

const gatherTokenRates = (updates: RateUpdate[]) =>
  updates.reduce((rates, update) => {
    const tokenId = toTokenId(update.id as WithTokenId)

    rates[tokenId] = {
      usd: {
        price: update.data.usd,
        change24hr: update.data.usd_24h_change
      }
    }

    return rates
  }, {} as Record<string, Record<string, Rate>>)

const handleNativeCurrencyUpdates = (updates: RateUpdate[]) => {
  log.debug('Handling native currency rate updates', { chains: updates.map((u) => u.id.chainId) })

  updates.forEach((u) => {
    storeApi.setNativeCurrencyRate(u.id.chainId, {
      price: u.data.usd,
      change24hr: u.data.usd_24h_change
    })

    const id = `native:${u.id.chainId}`
    const expiry = setTimeout(() => storeApi.removeNativeCurrencyRate(u.id.chainId), RATES_EXPIRY)

    clearTimeout(timeouts[id])
    timeouts[id] = expiry
  })
}

const handleTokenUpdates = (updates: RateUpdate[]) => {
  log.debug('Handling token rate updates', { addresses: updates.map((u) => u.id.address) })

  const rates = gatherTokenRates(updates)
  storeApi.setTokenRates(rates)

  updates.forEach((u) => {
    const address = u.id.address as string
    const expiry = setTimeout(() => storeApi.removeTokenRate(address), RATES_EXPIRY)

    clearTimeout(timeouts[address])
    timeouts[address] = expiry
  })
}

const handleUpdates = (rates: RateUpdate[]) => {
  const { token, native } = separateUpdates(rates)

  if (native.length) {
    handleNativeCurrencyUpdates(native)
  }

  if (token.length) {
    handleTokenUpdates(token)
  }
}

export { handleUpdates }
