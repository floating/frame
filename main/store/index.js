const Restore = require('react-restore')
const state = require('./state')
const actions = require('./actions')
const persist = require('./persist')

// TODO: Layer persisted op top of initial state

// const get = (path, obj = persist.get('main')) => {
//   path.split('.').some((key, i) => {
//     if (typeof obj !== 'object') { obj = undefined } else { obj = obj[key] }
//     return obj === undefined // Stop navigating the path if we get to undefined value
//   })
//   return obj
// }

// const persistedPaths = []

// persistedPaths.forEach(path => {
//   const value = get(path)
//   if (value !== undefined) store.__overwrite(path, value)
// })

const store = Restore.create(state(), actions)

store.observer(() => persist.set('main', store('main')))
module.exports = store
