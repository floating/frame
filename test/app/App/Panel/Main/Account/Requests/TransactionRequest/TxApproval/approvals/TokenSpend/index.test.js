import React from 'react'
import { fireEvent, render } from '@testing-library/react'
import { addHexPrefix } from 'ethereumjs-util'

import Restore from 'react-restore'
import store from '../../../../../../../../../../../main/store'
import ApproveTokenSpendComponent from '../../../../../../../../../../../app/App/Account/Account/Requests/TransactionRequest/TokenSpend'

jest.mock('../../../../../../../../../../../main/store/persist')

const TokenSpend = Restore.connect(ApproveTokenSpendComponent, store)

describe('changing approval amounts', () => {
  it('allows the user to set the token approval to a custom amount', async () => {
    return new Promise(async (resolve, reject) => {
      const requestedAmountHex = '0x011170'
      const approval = {
        id: 'erc20:approve',
        data: {
          spender: '0x9bc5baf874d2da8d216ae9f137804184ee5afef4',
          amount: requestedAmountHex,
          decimals: 4,
          name: 'TST',
          symbol: 'TST',
          spenderEns: '',
          spenderType: 'external',
          contract: '0x1eba19f260421142AD9Bf5ba193f6d4A0825e698'
        }
      }

      const { queryByRole, getByText } = render(
        <TokenSpend
          approval={approval}
          requestedAmountHex={requestedAmountHex}
          updateApproval={(amount) => {
            try {
              expect(amount).toBe('0x7a120')
              resolve()
            } catch (e) { 
              reject(e) 
            }
          }}
        />
      )

      const custom = queryByRole('button', { name: 'Custom' })
      fireEvent.click(custom)

      const enterAmount = queryByRole('textbox', { name: 'Custom Amount' })
      fireEvent.change(enterAmount, { target: { value: '50' } })

      const updateCustom = getByText('update')
      fireEvent.click(updateCustom)
    })
  })

  it('does not allows the user to set the token approval to a custom amount for an unknown token', () => {
    const requestedAmountHex = addHexPrefix(100e6.toString(16))

    const approval = {
      id: 'erc20:approve',
      data: {
        spender: '0x9bc5baf874d2da8d216ae9f137804184ee5afef4',
        amount: requestedAmountHex,
        decimals: 6,
        symbol: 'aUSDC',
        spenderEns: '',
        spenderType: 'external',
        contract: '0x1eba19f260421142AD9Bf5ba193f6d4A0825e698'
      }
    }

    const { queryByRole, getByText } = render(
      <TokenSpend
        approval={approval}
        requestedAmountHex={requestedAmountHex}
        updateApproval={() => {}}
      />
    )

    const custom = queryByRole('button', { name: 'Custom' })
    expect(custom).toBe(null)
  })

  it('allows the user to set the token approval to unlimited', async () => {
    return new Promise(async (resolve, reject) => {
      const requestedAmountHex = '0x011170'

      const approval = {
        id: 'erc20:approve',
        data: {
          spender: '0x9bc5baf874d2da8d216ae9f137804184ee5afef4',
          amount: requestedAmountHex,
          decimals: 4,
          name: 'TST',
          symbol: 'TST',
          spenderEns: '',
          spenderType: 'external',
          contract: '0x1eba19f260421142AD9Bf5ba193f6d4A0825e698'
        }
      }

      const { queryByRole, getByText } = render(
        <TokenSpend
          approval={approval}
          requestedAmountHex={requestedAmountHex}
          updateApproval={(amount) => {
            try {
              expect(amount).toBe('0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff')
              resolve()
            } catch (e) { 
              reject(e) 
            }
          }}
        />
      )

      const custom = queryByRole('button', { name: 'Custom' })
      fireEvent.click(custom)

      const setUnlimited = queryByRole('button', { name: 'Unlimited' })
      fireEvent.click(setUnlimited)
    })
  })

  it('allows the user to revert the token approval back to the original amount when no decimal data is present', async () => {
    return new Promise(async (resolve, reject) => {
      const requestedAmountHex = '0x011170'

      const approval = {
        id: 'erc20:approve',
        data: {
          spender: '0x9bc5baf874d2da8d216ae9f137804184ee5afef4',
          amount: requestedAmountHex,
          name: 'TST',
          symbol: 'TST',
          spenderEns: '',
          spenderType: 'external',
          contract: '0x1eba19f260421142AD9Bf5ba193f6d4A0825e698'
        }
      }

      const { queryByRole, getByText } = render(
        <TokenSpend
          approval={approval}
          requestedAmountHex={requestedAmountHex}
          updateApproval={(amount) => {
            try {
              if (amount === '0x011170') resolve()
            } catch (e) { 
              reject(e) 
            }
          }}
        />
      )

      const setUnlimited = queryByRole('button', { name: 'Unlimited' })
      fireEvent.click(setUnlimited)

      const setRequested = queryByRole('button', { name: 'Requested' })
      fireEvent.click(setRequested)
    })
  })

  it('allows the user to revert the token approval back to the original request', async () => {
    return new Promise(async (resolve, reject) => {
      const requestedAmountHex = '0x011170'

      const approval = {
        id: 'erc20:approve',
        data: {
          spender: '0x9bc5baf874d2da8d216ae9f137804184ee5afef4',
          amount: requestedAmountHex,
          decimals: 4,
          name: 'TST',
          symbol: 'TST',
          spenderEns: '',
          spenderType: 'external',
          contract: '0x1eba19f260421142AD9Bf5ba193f6d4A0825e698'
        }
      }

      const { queryByRole, getByText } = render(
        <TokenSpend
          approval={approval}
          requestedAmountHex={requestedAmountHex}
          updateApproval={(amount) => {
            try {
              if (amount === '0x011170') resolve()
            } catch (e) { 
              reject(e) 
            }
          }}
        />
      )

      const setUnlimited = queryByRole('button', { name: 'Unlimited' })
      fireEvent.click(setUnlimited)

      const setRequested = queryByRole('button', { name: 'Requested' })
      fireEvent.click(setRequested)
    })
  })

  const requiredApprovalData = ['decimals', 'symbol', 'name']

  requiredApprovalData.forEach(field => {
    it(`does not allow the user to edit the amount if ${field} is not present in approval data`, () => {
      const requestedAmountHex = addHexPrefix(100e6.toString(16))
      const approval = {
        id: 'erc20:approve',
        data: {
          spender: '0x9bc5baf874d2da8d216ae9f137804184ee5afef4',
          amount: requestedAmountHex,
          decimals: 6,
          name: 'TST',
          symbol: 'TST',
          spenderEns: '',
          spenderType: 'external',
          contract: '0x1eba19f260421142AD9Bf5ba193f6d4A0825e698'
        }
      }

      delete approval.data[field]

      const { queryByRole } = render(
        <TokenSpend
          approval={approval}
          requestedAmountHex={requestedAmountHex}
          updateApproval={() => {}}
        />
      )

      const custom = queryByRole('button', { name: 'Custom' })
      expect(custom).toBeNull()

      const requestedAmount = queryByRole('textbox')
      const displayedContent = requestedAmount.textContent.trim()
      expect(displayedContent).toBe(approval.data.decimals ? '100' : '100 million')

      // ensure click on requested amount textbox doesn't allow user to enter a custom amount
      fireEvent.click(requestedAmount)
      expect(queryByRole('textbox', { name: 'Custom Amount' })).toBeNull()
    })
  })
})

describe('formatting amounts', () => {
  const formattedAmounts = [
    { amount: 1e5, formatted: '100000' },
    { amount: 92e5, formatted: '9.2 million' },
    { amount: 100e9, formatted: '100 billion' },
    { amount: 2e12, formatted: '2 trillion' },
    { amount: 1e13, formatted: '~unlimited' }
  ]

  formattedAmounts.forEach(spec => {
    it(`formats a requested amount of ${spec.amount} as ${spec.formatted}`, () => {

      const amount = addHexPrefix((spec.amount * 1e6).toString(16))
      const requestedAmountHex = amount

      const approval = {
        id: 'erc20:approve',
        data: {
          spender: '0x9bc5baf874d2da8d216ae9f137804184ee5afef4',
          amount: amount,
          decimals: 6,
          name: 'TST',
          symbol: 'TST',
          spenderEns: '',
          spenderType: 'external',
          contract: '0x1eba19f260421142AD9Bf5ba193f6d4A0825e698'
        }
      }

      const { queryByRole, getByText } = render(
        <TokenSpend
          approval={approval}
          requestedAmountHex={requestedAmountHex}
          updateApproval={() => {}}
        />
      )

      const requestedAmount = queryByRole('textbox')
      const displayedContent = requestedAmount.textContent.trim()
      expect(displayedContent).toBe(spec.formatted)
    })
  })
})
