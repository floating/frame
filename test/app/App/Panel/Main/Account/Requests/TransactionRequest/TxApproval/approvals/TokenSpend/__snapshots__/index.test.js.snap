// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`revoking an approval should render the expected HTML 1`] = `
<div>
  <div
    class="approveTransactionWarning"
  >
    <div
      class="approveTransactionWarningOptions"
    >
      <div
        class="approveTransactionWarningReject"
        role="button"
      >
        Reject
      </div>
      <div
        class="approveTransactionWarningProceed"
        role="button"
      >
        Proceed
      </div>
    </div>
    <div
      class="approveTransactionWarningFill"
    >
      <div
        class="approveTransactionWarningIcon approveTransactionWarningIconLeft"
      >
        <div
          style="width: 32px; height: 32px;"
        >
          <svg
            height="32px"
            viewBox="0 0 24 24"
            width="32px"
          >
            <path
              d="M13 17.5a1 1 0 11-2 0 1 1 0 012 0zm-.25-8.25a.75.75 0 00-1.5 0v4.5a.75.75 0 001.5 0v-4.5z"
              fill="currentColor"
            />
            <path
              d="M9.836 3.244c.963-1.665 3.365-1.665 4.328 0l8.967 15.504c.963 1.667-.24 3.752-2.165 3.752H3.034c-1.926 0-3.128-2.085-2.165-3.752L9.836 3.244zm3.03.751a1 1 0 00-1.732 0L2.168 19.499A1 1 0 003.034 21h17.932a1 1 0 00.866-1.5L12.866 3.994z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="approveTransactionWarningIcon approveTransactionWarningIconRight"
      >
        <div
          style="width: 32px; height: 32px;"
        >
          <svg
            height="32px"
            viewBox="0 0 24 24"
            width="32px"
          >
            <path
              d="M13 17.5a1 1 0 11-2 0 1 1 0 012 0zm-.25-8.25a.75.75 0 00-1.5 0v4.5a.75.75 0 001.5 0v-4.5z"
              fill="currentColor"
            />
            <path
              d="M9.836 3.244c.963-1.665 3.365-1.665 4.328 0l8.967 15.504c.963 1.667-.24 3.752-2.165 3.752H3.034c-1.926 0-3.128-2.085-2.165-3.752L9.836 3.244zm3.03.751a1 1 0 00-1.732 0L2.168 19.499A1 1 0 003.034 21h17.932a1 1 0 00.866-1.5L12.866 3.994z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="approveTransactionWarningTitle"
      >
        revoke token approval
      </div>
      <div
        class="approveTokenSpend"
      >
        <div
          class="approveTokenSpendDescription"
        >
          <div
            class="approveTokenSpendSub"
          >
            wants to revoke approval to spend
          </div>
          <div
            class="approveTokenSpendToken"
          >
            <div
              class="approveTokenSpendTokenSymbol"
            >
              ???
            </div>
            <div
              class="approveTokenSpendTokenContract"
            />
          </div>
          <div
            class="approveTokenSpendTokenName"
          >
            Unknown Token
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render the expected HTML 1`] = `
<div>
  <div
    class="approveTransactionWarning"
  >
    <div
      class="approveTransactionWarningOptions"
    >
      <div
        class="approveTransactionWarningReject"
        role="button"
      >
        Reject
      </div>
      <div
        class="approveTokenSpendEditButton"
        role="button"
      >
        Edit
      </div>
      <div
        class="approveTransactionWarningProceed"
        role="button"
      >
        Proceed
      </div>
    </div>
    <div
      class="approveTransactionWarningFill"
    >
      <div
        class="approveTransactionWarningIcon approveTransactionWarningIconLeft"
      >
        <div
          style="width: 32px; height: 32px;"
        >
          <svg
            height="32px"
            viewBox="0 0 24 24"
            width="32px"
          >
            <path
              d="M13 17.5a1 1 0 11-2 0 1 1 0 012 0zm-.25-8.25a.75.75 0 00-1.5 0v4.5a.75.75 0 001.5 0v-4.5z"
              fill="currentColor"
            />
            <path
              d="M9.836 3.244c.963-1.665 3.365-1.665 4.328 0l8.967 15.504c.963 1.667-.24 3.752-2.165 3.752H3.034c-1.926 0-3.128-2.085-2.165-3.752L9.836 3.244zm3.03.751a1 1 0 00-1.732 0L2.168 19.499A1 1 0 003.034 21h17.932a1 1 0 00.866-1.5L12.866 3.994z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="approveTransactionWarningIcon approveTransactionWarningIconRight"
      >
        <div
          style="width: 32px; height: 32px;"
        >
          <svg
            height="32px"
            viewBox="0 0 24 24"
            width="32px"
          >
            <path
              d="M13 17.5a1 1 0 11-2 0 1 1 0 012 0zm-.25-8.25a.75.75 0 00-1.5 0v4.5a.75.75 0 001.5 0v-4.5z"
              fill="currentColor"
            />
            <path
              d="M9.836 3.244c.963-1.665 3.365-1.665 4.328 0l8.967 15.504c.963 1.667-.24 3.752-2.165 3.752H3.034c-1.926 0-3.128-2.085-2.165-3.752L9.836 3.244zm3.03.751a1 1 0 00-1.732 0L2.168 19.499A1 1 0 003.034 21h17.932a1 1 0 00.866-1.5L12.866 3.994z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="approveTransactionWarningTitle"
      >
        token approval
      </div>
      <div
        class="approveTokenSpend"
      >
        <div
          class="approveTokenSpendDescription"
        >
          <div
            class="approveTokenSpendSub"
          >
            wants approval to spend
          </div>
          <div
            class="approveTokenSpendToken"
          >
            <div
              class="approveTokenSpendTokenSymbol"
            >
              ???
            </div>
            <div
              class="approveTokenSpendTokenContract"
            />
          </div>
          <div
            class="approveTokenSpendTokenName"
          >
            Unknown Token
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
